File: java/roman-numbers/src/test/java/ArabicToRomanTest.java
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class ArabicToRomanTest {

    private RomanConverter converter;

    @BeforeEach
    void setUp() {
        converter = new RomanConverter();
    }

    @Test
    void it_should_convert_1s_to_Is() {
        assertThat(converter.toRoman(1)).isEqualTo("I");
        assertThat(converter.toRoman(2)).isEqualTo("II");
        assertThat(converter.toRoman(3)).isEqualTo("III");
    }

    @Test
    void it_should_convert_5_to_V() {
        assertThat(converter.toRoman(5)).isEqualTo("V");
        assertThat(converter.toRoman(6)).isEqualTo("VI");
        assertThat(converter.toRoman(7)).isEqualTo("VII");
    }

    @Test
    void it_should_convert_10s_to_Xs() {
        assertThat(converter.toRoman(10)).isEqualTo("X");
        assertThat(converter.toRoman(18)).isEqualTo("XVIII");
        assertThat(converter.toRoman(20)).isEqualTo("XX");
        assertThat(converter.toRoman(36)).isEqualTo("XXXVI");
    }

    @Test
    void it_should_convert_according_to_subtraction_rules() {
        assertThat(converter.toRoman(4)).isEqualTo("IV");
        assertThat(converter.toRoman(9)).isEqualTo("IX");
        assertThat(converter.toRoman(40)).isEqualTo("XL");
        assertThat(converter.toRoman(90)).isEqualTo("XC");
        assertThat(converter.toRoman(400)).isEqualTo("CD");
        assertThat(converter.toRoman(900)).isEqualTo("CM");
    }

    @Test
    void it_should_convert_50_to_L() {
        assertThat(converter.toRoman(50)).isEqualTo("L");
        assertThat(converter.toRoman(73)).isEqualTo("LXXIII");
    }

    @Test
    void it_should_convert_100s_to_Cs() {
        assertThat(converter.toRoman(100)).isEqualTo("C");
        assertThat(converter.toRoman(123)).isEqualTo("CXXIII");
    }

    @Test
    void it_should_convert_500_to_D() {
        assertThat(converter.toRoman(500)).isEqualTo("D");
        assertThat(converter.toRoman(857)).isEqualTo("DCCCLVII");
    }

    @Test
    void it_should_convert_1000s_to_Ms() {
        assertThat(converter.toRoman(1000)).isEqualTo("M");
        assertThat(converter.toRoman(1978)).isEqualTo("MCMLXXVIII");
    }
}

File: kotlin/roman-numbers-reverse/src/test/kotlin/ArabicToRomanTest.kt
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class `arabic to roman conversion` {
    @Nested
    inner class `summation rules` {
        @Test
        fun `convert zero should return empty string`() {
            assertThat(Arabic(0).toRoman()).isEqualTo("")
        }

        @Test
        fun `convert should return I's for every number less than 4`() {
            assertThat(Arabic(1).toRoman()).isEqualTo("I")
            assertThat(Arabic(2).toRoman()).isEqualTo("II")
            assertThat(Arabic(3).toRoman()).isEqualTo("III")
        }

        @Test
        fun `convert should return V's for every number from 5 up to 9`() {
            assertThat(Arabic(5).toRoman()).isEqualTo("V")
            assertThat(Arabic(6).toRoman()).isEqualTo("VI")
            assertThat(Arabic(8).toRoman()).isEqualTo("VIII")
        }

        @Test
        fun `convert should return X's for every number from 10 up to 40`() {
            assertThat(Arabic(10).toRoman()).isEqualTo("X")
            assertThat(Arabic(35).toRoman()).isEqualTo("XXXV")
        }

        @Test
        fun `convert should return L's for every number from 50 up to 90`() {
            assertThat(Arabic(50).toRoman()).isEqualTo("L")
            assertThat(Arabic(73).toRoman()).isEqualTo("LXXIII")
        }

        @Test
        fun `convert should return C's for every number from 100 up to 400`() {
            assertThat(Arabic(100).toRoman()).isEqualTo("C")
            assertThat(Arabic(352).toRoman()).isEqualTo("CCCLII")
        }

        @Test
        fun `convert should return D's for every number from 500 up to 900`() {
            assertThat(Arabic(500).toRoman()).isEqualTo("D")
            assertThat(Arabic(666).toRoman()).isEqualTo("DCLXVI")
        }

        @Test
        fun `convert should return M's for every number from 1000`() {
            assertThat(Arabic(1000).toRoman()).isEqualTo("M")
            assertThat(Arabic(1236).toRoman()).isEqualTo("MCCXXXVI")
        }
    }
    
    @Nested
    inner class `subtraction rules` {
        @Test
        fun `convert should return IX for 9`() {
            assertThat(Arabic(9).toRoman()).isEqualTo("IX")
        }

        @Test
        fun `convert should return IV for 4`() {
            assertThat(Arabic(4).toRoman()).isEqualTo("IV")
        }

        @Test
        fun `convert should return XL for 40`() {
            assertThat(Arabic(40).toRoman()).isEqualTo("XL")
        }

        @Test
        fun `convert should return XC for 90`() {
            assertThat(Arabic(90).toRoman()).isEqualTo("XC")
        }

        @Test
        fun `convert should return CD for 400`() {
            assertThat(Arabic(400).toRoman()).isEqualTo("CD")
        }

        @Test
        fun `convert should return CM for 900`() {
            assertThat(Arabic(900).toRoman()).isEqualTo("CM")
        }
    }
    
    @Nested
    inner class `high level tests` {
        @Test
        fun `it should convert 1978`() {
            assertThat(Arabic(1978).toRoman()).isEqualTo("MCMLXXVIII")
        }

        @Test
        fun `it should convert 1984`() {
            assertThat(Arabic(1984).toRoman()).isEqualTo("MCMLXXXIV")
        }
    }
}